### PR TITLE
Add gpt2 implementation

### PR DESCRIPTION
**Pull request by aliciafmachado**
Created a new file called "gpt2.ts" which contains an implementation of GPT2. I followed the implementation from Hugging Face in this [file](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_tf_gpt2.py).

This fixes the issues #31 and #49. Loading GPT2 weights will be added on a follow-up PR. 

The main differences from the previous implementation are around the layer norm, dropout and residual connection orders in the transformer computation. 

I also had to extend the layer norm to have an array of parameters instead of single values of gain and bias. Also removed epsilon as a trainable parameter which was likely causing layer norm to not work properly.

Summary of changes:
1. Added GPT2 implementation and tests on checking number of parameters and positional encodings.
2. Extended layer norm implementation, and fixed bug with epsilon being a parameter.

### CODE REVIEW

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+ */
+export function lastTokenCrossEntropyLoss(
+  model: {
+    params: { tokenEmbedding: GTensor<'tokenId' | 'inputRep'> };
+  },
+  computation: TransformerComputation,
+  targetTokenIdxs: GTensor<'batch'>
+): tf.Scalar {
+  const logits = lastTokenLogits(model, computation);
+  const logProbs = logits.softmax('tokenId').log();
```
Comment (iislucas) [inline animated-transformer/src/lib/transformer/gpt2.ts:454]: probably want to drop that log statement now.

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+  const oneHotToken = new GTensor(oneHot(targetTokenIdxs.tensor, nTokens), ['batch', 'tokenId']);
+  const crossEntopy = logProbs.pointwiseMul(oneHotToken);
+  return (
+    crossEntopy
+      .sumOverDims(['batch', 'tokenId'])
+      // ._tfScalarMul(tf.scalar(-1))
+      ._tfScalarDiv(tf.scalar(targetTokenIdxs.dim.batch.size * -1)).tensor as tf.Scalar
+  );
+  // const squaredError = signedDelta.pointwiseMul(signedDelta);
+  // const loss = squaredError.sumOverDims(['batch', 'token']);
```
Comment (iislucas) [inline animated-transformer/src/lib/transformer/gpt2.ts:465]: remove comments?

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+  targetTokenIdxs: GTensor<'batch'>
+): tf.Scalar {
+  const predictions = transformerTopPrediction(model, computation);
+  return predictions
+    .pointwiseEqual(targetTokenIdxs)
+    .sumOverDims(['batch'])
+    .tensor.div(tf.scalar(targetTokenIdxs.dim.batch.size));
+}
+
+export function addPosEmbeddings (
```
Comment (iislucas) [inline animated-transformer/src/lib/transformer/gpt2.ts:499]: indentation is odd within this function def.

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+    let finalOutput = lastBlock;
+    finalOutput.seqOutput = finalSeqOutput;
+    finalOutput.seqInput = lastBlock.seqOutput;
+    compute.layers.push(finalOutput);
+  }
+  
+  return compute;
+}
+
+/** Batch compute the loss for the last token of a transformer.
```
Comment (iislucas) [inline animated-transformer/src/lib/transformer/gpt2.ts:424]: I think the comment here is wrong; this doesn't compute the loss but the final tokenId's for each example in the batch.

**Review by iislucas — COMMENTED** [review]
Thanks! Great to have this! A few comments, mostly just minor cleanups.

File: animated-transformer/src/lib/gtensor/layer_norm.ts [hunk 5]
```diff
@@ -47,10 +65,10 @@ export function layerNorm<G extends string, D extends G>(
     .scalarDiv(repSizeScalar);
   const meanAndVarNormalized = g
     .pointwiseSub(mean)
-    .pointwiseDiv(varianceSquared.scalarAdd(epsilon).sqrt());
-  const scaledNorm = meanAndVarNormalized.scalarMul(gain);
+    .pointwiseDiv(varianceSquared.pointwiseAdd(epsilon).sqrt());
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/gtensor/layer_norm.ts:68]: It's true. Epsilon is a scalar so scalarAdd is enough. But for bias and gain, we need pointwise operations.

File: animated-transformer/src/lib/gtensor/layer_norm.ts [hunk 6]
```diff
@@ -13,31 +13,49 @@ See the License for the specific language governing permissions and
...
+  }
+  return layerNormParams;
+}
+
+export function layerNorm<G extends string, D extends G, T extends G = never>(
+  layerNormParams: LayerNormParams<T>,
   g: GTensor<G>,
-  dim: D
+  dim: D,
+  eps: number = 1e-5
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/gtensor/layer_norm.ts:55]: Made it a gTensor, but I think that we will need to reinit the epsilon, unless we store it as a gtensor in the compute spec. What are your thoughts on it?

File: animated-transformer/src/lib/gtensor/layer_norm.spec.ts [hunk 7]
```diff
@@ -55,11 +54,61 @@ describe('layer_norm', () => {
...
+    const varSqrd3 = ((2 - 6) ^ (2 + (4 - 6)) ^ (2 + (12 - 6)) ^ 2) / 3;
+    const approxStdDev3 = Math.sqrt(varSqrd3 + epsilonNum);
+
+    const gNormed = layerNorm({ gain, bias }, g, 'rep', epsilonNum);
+
+    tf.test_util.expectArraysClose(gNormed.transposeTo(['pos', 'rep']).tensor.dataSync(), [
+      [0, 0, 0],
+      [-1, 0, 1].map((x) => x / Math.sqrt(2 + 1e3)).map((x) => x + 1),
+      [-4, -2, 6].map((x) => x / approxStdDev3).map((x) => x + 2),
+    ]);
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/gtensor/layer_norm.spec.ts:84]: Thanks :), just followed the pattern :D

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+  };
+
+// ---------------------------------------------------------------------------
+
+export type AttnHeadParamSpec = {
+  inputRep: number;
+  kq: number;
+  heads: number;
+  value: number;
+  // ffRep: number;
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:81]: Removing the commented lines. I just copied what was in `transformer_gtensor.ts` but i cleaned-up all lines that are not meaningful in this context.

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+  dropoutRate: number;
+  layerNormEpsilon: number;
+};
+
+// Default GPT2 config for eval.
+// Dropout is disabled below. For training it should be 0.1.
+export function defaultGPT2EvalConfig(
+    tokenRep: BasicTaskTokenRep
+): TransformerConfig {
+  const embedding_size = 768;
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:111]: Done.

File: animated-transformer/src/lib/transformer/gpt2.ts (hunk 1 above)
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:454]: Do you mean the logProbs? Why?

File: animated-transformer/src/lib/transformer/gpt2.ts (hunk 2 above)
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:465]: Done.

File: animated-transformer/src/lib/transformer/gpt2.ts (hunk 3 above)
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:499]: Fixed it!

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+  return examplePredictions;
+}
+
+export function makeTransformer(transformerConfig: TransformerConfig): TransformerModel {
+  const config = structuredClone(transformerConfig);
+  const params = initDecoderParams(config);
+  return { config, params };
+}
+
+// export const transformerModelKind = modelRegistry.register(
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:571]: same as before! they came when i forked the file. Cleaned them up!

File: animated-transformer/src/lib/transformer/gpt2.ts (hunk 4 above)
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:424]: Same as before (came with the file). Fixed the comment to say the right thing!

File: animated-transformer/src/lib/transformer/gpt2.spec.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+==============================================================================*/
+
+import { GTensor } from '../gtensor/gtensor';
+import * as tf from '@tensorflow/tfjs';
+import * as transformer from './gpt2';
+import { prepareBasicTaskTokenRep, BasicTaskTokenRep } from '../tokens/token_gemb';
+import * as jstree from '../js_tree/js_tree';
+
+function generateTestTask(): BasicTaskTokenRep {
+  const tokens = Array(50257 - 4).fill("test");
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.spec.ts:23]: Done!

File: animated-transformer/src/lib/transformer/gpt2.spec.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+      params
+    );
+
+    // Check full GPT2 size.
+    expect(paramCountGPT2).toEqual(124439808);
+  });
+
+  it('Test positional embeddings.', async () => {
+    // Set dummy transformer for testing.
+    const embedding_size = 2;
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.spec.ts:62]: Fixed it!

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+} from '../gtensor/params';
+
+import * as tf_init from '@tensorflow/tfjs-layers/dist/initializers';
+import { initLayerNormParamsWithDims, layerNorm, LayerNormParams } from '../gtensor/layer_norm';
+import { dropout } from './dropout';
+import { BasicTaskTokenRep, StrSeqPrepFn } from '../tokens/token_gemb';
+import { RandomStream } from '../random/random';
+
+// ---------------------------------------------------------------------------
+export type TransformerConfig = {
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:40]: I believe that given that we pass a `kind` field under this struct, perhaps it should be just "Config" as you suggested. Changed it accordingly.

Do you think we should do the same for other functions / params that start with Transformer?

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 14]
```diff
@@ -0,0 +1,574 @@
...
+
+export type AttnHeadComputeSpec = {
+  // Whether to include or not the residual connections in the computation.
+  residuals: boolean;
+  dropoutRate: number;
+  layerNormEpsilon: number;
+};
+
+// Default GPT2 config for eval.
+// Dropout is disabled below. For training it should be 0.1.
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:107]: Agree, added a "training" flag.

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+  b: GTensor<Output>;
+};
+
+// Use of type here to be compatible with generic params.
+export type AttnHeadParams = {
+  queryM: GTensor<'heads' | 'inputRep' | 'kq'>;
+  keyM: GTensor<'heads' | 'inputRep' | 'kq'>;
+  valueM: GTensor<'heads' | 'inputRep' | 'value'>;
+  headsToInputRepM: GTensor<'heads' | 'value' | 'inputRepToFF'>;
+  // workaround for https://github.com/microsoft/TypeScript/issues/48070
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:160]: Same as the other comments. Cleaned it up!

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 16]
```diff
@@ -0,0 +1,574 @@
...
+  tokenEmbedding: GTensor<'tokenId' | 'inputRep'>;
+  posEmbedding?: GTensor<'posId' | 'inputRep'>;
+  layerNorm?: LayerNormParams<'inputRep'>; // 768 + 768
+};
+
+export type VarTransformerParams = VarifyTensorParams<TransformerParams>;
+export type SerialTransformerParams = SerializeTensorParams<TransformerParams>;
+
+export type TransformerModel = {
+  // Locally cached version of the model.
```
[the remark this answers is not in the document]
Comment (aliciafmachado) [reply animated-transformer/src/lib/transformer/gpt2.ts:184]: Same as before! Cleaned it up!

**Comment by aliciafmachado** [comment]
Fixed what you mentioned in the comments. Feel free to take a second look @iislucas !

I think that the main learning here is that I should have cleaned the file when I forked from transformer_gtensor.py. Perhaps we should also clean it up, either here or on a follow-up CL to not mix things up.

File: animated-transformer/src/lib/transformer/gpt2.ts [hunk 17]
```diff
@@ -0,0 +1,567 @@
...
+      model.config.spec.layers[i].computeSpec,
+      layerParams,
+      currentLayerInput,
+      generator
+    );
+    compute.layers.push(layerCompute);
+    currentLayerInput = layerCompute.seqOutput;
+  });
+
+  // TODO(@aliciafmachado): Hacky way to apply layer norm after the final layer.
```
Comment (aliciafmachado) [inline animated-transformer/src/lib/transformer/gpt2.ts:405]: Any thoughts on this? Not sure if there would be a easy way to hook this in, and we have been only using the last layer in compute for now anyway.

File: animated-transformer/src/lib/gtensor/layer_norm.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (iislucas) [reply animated-transformer/src/lib/gtensor/layer_norm.ts]: gah! I actually need to think more about this, maybe best to put it back as it was actually. The bigger question is about how to handle constants in a better way, I have some ideas, but it'll be a different pull request for sure.

File: animated-transformer/src/lib/transformer/gpt2.ts (hunk 17 above)
Comment (iislucas) [reply animated-transformer/src/lib/transformer/gpt2.ts:405]: I'm not sure; but I think it's ok, definitely ok for now.

**Review by iislucas — APPROVED** [review]

File: animated-transformer/src/lib/gtensor/layer_norm.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (iislucas) [reply animated-transformer/src/lib/gtensor/layer_norm.ts]: actually, reading in more detail, I think what you've done is fine (and a little better than before) but I will still want to have a good hard think about constants. Thanks!

**Comment by iislucas** [comment]
Thanks! Looks good, merge at will!

**Comment by aliciafmachado** [comment]
Sounds good! Will merge! 

Just an observation: I just added a commit to fix a typo in transformer_gtensor.ts and reuse the BatchAttnHeadComputation from transformer_gtensor.ts in gpt2.ts as it is the same and it will make it easier to use the functions Laura implemented without having to copy paste into this gpt2.ts.

**Comment by iislucas** [comment]
Thanks! That's great!
